src/core/storage/client.ts

```diff
@@ -0,0 +1,179 @@
+      try {
+        const store = new StoreMemory();
+        this.storage = await Storage.create({ principal: this.config.signer, store });
+        await this.storage.addSpace(this.config.delegation);
+      } catch (error: unknown) {
+        const message = error instanceof Error ? error.message : 'Unknown error';
+        throw new Error(`Failed to initialize storage client: ${message}`, { cause: error });
+      }
+    })();
+
+    return this.initialized;
+  }
+
+  /**
+   * Get the storage client
+   * @returns The storage client
+   */
+  getStorage(): Storage.Client | null {
+    return this.storage;
+  }
+
+  /**
+   * Check if the client is connected and ready
+   */
+  isConnected(): boolean {
+    return this.initialized !== null;
+  }
+
+  /**
+   * Get the storage config
+   * @returns The storage config
+   */
+  getConfig(): StorageConfig {
+    return this.config;
+  }
+
+  /**
+   * Get the configured gateway URL
+   * @returns The configured gateway URL
+   * @throws Error if gateway URL is not set
+   */
+  getGatewayUrl(): URL {
+    if (!this.config.gatewayUrl) {
+      throw new Error('Gateway URL is not set');
+    }
+    return this.config.gatewayUrl;
+  }
+
+  /**
+   * Upload files to Storacha network
+   * The Storage Client needs to be initialized to upload files.
+   *
+   * @param files - Array of files to upload
+   * @param options - Upload options
+   * @returns The uploaded files' Content ID and URL
+   */
+  async uploadFiles(files: UploadFile[], options: UploadOptions = {}): Promise<UploadResult> {
+    if (!this.initialized || !this.storage) {
+      throw new Error('Client not initialized');
+    }
+
+    try {
+      if (options.signal?.aborted) {
+        throw new Error('Upload aborted');
+      }
+
+      const fileObjects = files.map(file => {
+        const buffer = Buffer.from(file.content, 'base64');
+        return new File([buffer], file.name, {
+          type: file.type,
+        });
+      });
+
+      // TODO: Use uploadDirectory when we implement the change to collect and return CIDs of each file, and fix the retrieve tool to use paths instead of CIDs.
+      // const root = await this.storage.uploadDirectory(fileObjects, {
+      //   // If publishToFilecoin is false, we don't provide a pieceHasher, so the content is not pinned to the Filecoin Network
+      //   ...(options.publishToFilecoin === false ? { pieceHasher: undefined } : {}),
+      //   retries: options.retries ?? 3,
+      //   signal: options.signal,
+      // });
+
+      // TODO: Use uploadFile when we implement the change to collect and return CIDs of each file, and fix the retrieve tool to use paths instead of CIDs.
+      const root = await this.storage.uploadFile(fileObjects[0], {
+        // If publishToFilecoin is false, we don't provide a pieceHasher, so the content is not pinned to the Filecoin Network
+        ...(options.publishToFilecoin === false ? { pieceHasher: undefined } : {}),
```

Comment:
I thought this was by default `false`?

```suggestion
        ...(options.publishToFilecoin === true ? {} : { pieceHasher: undefined }),
```

Worth checking if `undefined` works here - I was thinking you'd need to pass `null`.

Reply:
You are right. I've changed the default logic as you suggested, and regarding the `undefined` it only accepts `undefined`, `null` is not supported.